### PR TITLE
r/apigateway_domain_name - support for private endpoints

### PR DESCRIPTION
**Pull request by DrFaust92**
### Relations

Relates #40274 - there are at least 2 more PRs needed for different resources: 1. base mapping support and 2. domain name association with VPC (new resource)

### Output from Acceptance Testing

```console
% make testacc TESTS=TestAccAPIGatewayDomainName_  PKG=apigateway

--- PASS: TestAccAPIGatewayDomainName_privatePolicy (15.84s)
--- PASS: TestAccAPIGatewayDomainName_private (15.84s)
--- PASS: TestAccAPIGatewayDomainName_securityPolicy (19.83s)
```

### CODE REVIEW

**Review by ewbankkit — DISMISSED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccAPIGatewayDomainName_\|TestAccAPIGatewayDomainNameDataSource_' PKG=apigateway ACCTEST_PARALLELISM=3
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.3 test ./internal/service/apigateway/... -v -count 1 -parallel 3  -run=TestAccAPIGatewayDomainName_\|TestAccAPIGatewayDomainNameDataSource_ -timeout 360m
2024/12/04 15:44:39 Initializing Terraform AWS Provider...
=== RUN   TestAccAPIGatewayDomainNameDataSource_tags
=== PAUSE TestAccAPIGatewayDomainNameDataSource_tags
=== RUN   TestAccAPIGatewayDomainNameDataSource_tags_NullMap
=== PAUSE TestAccAPIGatewayDomainNameDataSource_tags_NullMap
=== RUN   TestAccAPIGatewayDomainNameDataSource_tags_EmptyMap
=== PAUSE TestAccAPIGatewayDomainNameDataSource_tags_EmptyMap
=== RUN   TestAccAPIGatewayDomainNameDataSource_tags_DefaultTags_nonOverlapping
=== PAUSE TestAccAPIGatewayDomainNameDataSource_tags_DefaultTags_nonOverlapping
=== RUN   TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_DefaultTag
=== PAUSE TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_DefaultTag
=== RUN   TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_ResourceTag
=== PAUSE TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_ResourceTag
=== RUN   TestAccAPIGatewayDomainNameDataSource_basic
=== PAUSE TestAccAPIGatewayDomainNameDataSource_basic
=== RUN   TestAccAPIGatewayDomainName_tags
=== PAUSE TestAccAPIGatewayDomainName_tags
=== RUN   TestAccAPIGatewayDomainName_tags_null
=== PAUSE TestAccAPIGatewayDomainName_tags_null
=== RUN   TestAccAPIGatewayDomainName_tags_EmptyMap
=== PAUSE TestAccAPIGatewayDomainName_tags_EmptyMap
=== RUN   TestAccAPIGatewayDomainName_tags_AddOnUpdate
=== PAUSE TestAccAPIGatewayDomainName_tags_AddOnUpdate
=== RUN   TestAccAPIGatewayDomainName_tags_EmptyTag_OnCreate
=== PAUSE TestAccAPIGatewayDomainName_tags_EmptyTag_OnCreate
=== RUN   TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Add
=== PAUSE TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Add
=== RUN   TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Replace
=== PAUSE TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Replace
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_providerOnly
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_providerOnly
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_nonOverlapping
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_nonOverlapping
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_overlapping
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_overlapping
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_updateToProviderOnly
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_updateToProviderOnly
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_updateToResourceOnly
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_updateToResourceOnly
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_emptyResourceTag
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_emptyResourceTag
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_emptyProviderOnlyTag
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_emptyProviderOnlyTag
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_nullOverlappingResourceTag
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_nullOverlappingResourceTag
=== RUN   TestAccAPIGatewayDomainName_tags_DefaultTags_nullNonOverlappingResourceTag
=== PAUSE TestAccAPIGatewayDomainName_tags_DefaultTags_nullNonOverlappingResourceTag
=== RUN   TestAccAPIGatewayDomainName_tags_ComputedTag_OnCreate
=== PAUSE TestAccAPIGatewayDomainName_tags_ComputedTag_OnCreate
=== RUN   TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Add
=== PAUSE TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Add
=== RUN   TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Replace
=== PAUSE TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Replace
=== RUN   TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_DefaultTag
=== PAUSE TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_DefaultTag
=== RUN   TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_ResourceTag
=== PAUSE TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_ResourceTag
=== RUN   TestAccAPIGatewayDomainName_certificateARN
    domain_name_test.go:29: Environment variable ACM_CERTIFICATE_ROOT_DOMAIN is not set. For DNS validation requests, this domain must be publicly accessible and configurable via Route53 during the testing. For email validation requests, you must have access to one of the five standard email addresses used (admin|administrator|hostmaster|postmaster|webmaster)@domain or one of the WHOIS contact addresses.
--- SKIP: TestAccAPIGatewayDomainName_certificateARN (0.00s)
=== RUN   TestAccAPIGatewayDomainName_certificateName
    domain_name_test.go:65: Environment variable AWS_API_GATEWAY_DOMAIN_NAME_CERTIFICATE_BODY is not set. This environment variable must be set to any non-empty value with a publicly trusted certificate body to enable the test.
--- SKIP: TestAccAPIGatewayDomainName_certificateName (0.00s)
=== RUN   TestAccAPIGatewayDomainName_regionalCertificateARN
=== PAUSE TestAccAPIGatewayDomainName_regionalCertificateARN
=== RUN   TestAccAPIGatewayDomainName_regionalCertificateName
    domain_name_test.go:161: Environment variable AWS_API_GATEWAY_DOMAIN_NAME_REGIONAL_CERTIFICATE_NAME_ENABLED is not set. This environment variable must be set to any non-empty value in a region where uploading REGIONAL certificates is allowed to enable the test.
--- SKIP: TestAccAPIGatewayDomainName_regionalCertificateName (0.00s)
=== RUN   TestAccAPIGatewayDomainName_securityPolicy
=== PAUSE TestAccAPIGatewayDomainName_securityPolicy
=== RUN   TestAccAPIGatewayDomainName_private
=== PAUSE TestAccAPIGatewayDomainName_private
=== RUN   TestAccAPIGatewayDomainName_privatePolicy
=== PAUSE TestAccAPIGatewayDomainName_privatePolicy
=== RUN   TestAccAPIGatewayDomainName_disappears
=== PAUSE TestAccAPIGatewayDomainName_disappears
=== RUN   TestAccAPIGatewayDomainName_MutualTLSAuthentication_basic
    domain_name_test.go:321: Environment variable ACM_CERTIFICATE_ROOT_DOMAIN is not set. For DNS validation requests, this domain must be publicly accessible and configurable via Route53 during the testing. For email validation requests, you must have access to one of the five standard email addresses used (admin|administrator|hostmaster|postmaster|webmaster)@domain or one of the WHOIS contact addresses.
--- SKIP: TestAccAPIGatewayDomainName_MutualTLSAuthentication_basic (0.00s)
=== RUN   TestAccAPIGatewayDomainName_MutualTLSAuthentication_ownership
    domain_name_test.go:377: Environment variable ACM_CERTIFICATE_ROOT_DOMAIN is not set. For DNS validation requests, this domain must be publicly accessible and configurable via Route53 during the testing. For email validation requests, you must have access to one of the five standard email addresses used (admin|administrator|hostmaster|postmaster|webmaster)@domain or one of the WHOIS contact addresses.
--- SKIP: TestAccAPIGatewayDomainName_MutualTLSAuthentication_ownership (0.00s)
=== RUN   TestAccAPIGatewayDomainName_updateIDFormat
=== PAUSE TestAccAPIGatewayDomainName_updateIDFormat
=== CONT  TestAccAPIGatewayDomainNameDataSource_tags
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_updateToProviderOnly
=== CONT  TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_DefaultTag
--- PASS: TestAccAPIGatewayDomainNameDataSource_tags (19.46s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_nullNonOverlappingResourceTag
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_nullNonOverlappingResourceTag (56.85s)
=== CONT  TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Replace
--- PASS: TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Replace (68.51s)
=== CONT  TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Add
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_updateToProviderOnly (159.88s)
=== CONT  TestAccAPIGatewayDomainName_tags_ComputedTag_OnCreate
--- PASS: TestAccAPIGatewayDomainName_tags_ComputedTag_OnUpdate_Add (48.43s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_emptyProviderOnlyTag
--- PASS: TestAccAPIGatewayDomainName_tags_ComputedTag_OnCreate (89.68s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_nullOverlappingResourceTag
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_emptyProviderOnlyTag (70.13s)
=== CONT  TestAccAPIGatewayDomainName_tags_EmptyMap
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_nullOverlappingResourceTag (22.22s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_overlapping
--- PASS: TestAccAPIGatewayDomainName_tags_EmptyMap (84.64s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_nonOverlapping
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_overlapping (106.15s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_providerOnly
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_nonOverlapping (60.98s)
=== CONT  TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Replace
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_providerOnly (94.74s)
=== CONT  TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Add
--- PASS: TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Replace (85.04s)
=== CONT  TestAccAPIGatewayDomainName_tags_EmptyTag_OnCreate
--- PASS: TestAccAPIGatewayDomainName_tags_EmptyTag_OnUpdate_Add (66.82s)
=== CONT  TestAccAPIGatewayDomainName_tags_AddOnUpdate
--- PASS: TestAccAPIGatewayDomainName_tags_EmptyTag_OnCreate (68.96s)
=== CONT  TestAccAPIGatewayDomainName_private
--- PASS: TestAccAPIGatewayDomainName_tags_AddOnUpdate (44.78s)
=== CONT  TestAccAPIGatewayDomainName_updateIDFormat
--- PASS: TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_DefaultTag (613.03s)
=== CONT  TestAccAPIGatewayDomainName_disappears
--- PASS: TestAccAPIGatewayDomainName_updateIDFormat (91.55s)
=== CONT  TestAccAPIGatewayDomainName_privatePolicy
--- PASS: TestAccAPIGatewayDomainName_private (132.54s)
=== CONT  TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_ResourceTag
--- PASS: TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_ResourceTag (24.30s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_emptyResourceTag
--- PASS: TestAccAPIGatewayDomainName_privatePolicy (97.26s)
=== CONT  TestAccAPIGatewayDomainName_tags_null
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_emptyResourceTag (80.52s)
=== CONT  TestAccAPIGatewayDomainName_tags
--- PASS: TestAccAPIGatewayDomainName_tags_null (84.52s)
=== CONT  TestAccAPIGatewayDomainName_tags_DefaultTags_updateToResourceOnly
--- PASS: TestAccAPIGatewayDomainName_disappears (254.09s)
=== CONT  TestAccAPIGatewayDomainNameDataSource_basic
--- PASS: TestAccAPIGatewayDomainName_tags (90.83s)
=== CONT  TestAccAPIGatewayDomainNameDataSource_tags_DefaultTags_nonOverlapping
--- PASS: TestAccAPIGatewayDomainNameDataSource_tags_DefaultTags_nonOverlapping (19.39s)
=== CONT  TestAccAPIGatewayDomainNameDataSource_tags_EmptyMap
--- PASS: TestAccAPIGatewayDomainNameDataSource_basic (89.48s)
=== CONT  TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_DefaultTag
--- PASS: TestAccAPIGatewayDomainName_tags_DefaultTags_updateToResourceOnly (121.19s)
=== CONT  TestAccAPIGatewayDomainName_regionalCertificateARN
--- PASS: TestAccAPIGatewayDomainNameDataSource_tags_EmptyMap (109.29s)
=== CONT  TestAccAPIGatewayDomainName_securityPolicy
--- PASS: TestAccAPIGatewayDomainName_regionalCertificateARN (67.34s)
=== CONT  TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_ResourceTag
--- PASS: TestAccAPIGatewayDomainName_securityPolicy (75.23s)
=== CONT  TestAccAPIGatewayDomainNameDataSource_tags_NullMap
--- PASS: TestAccAPIGatewayDomainNameDataSource_tags_NullMap (18.97s)
--- PASS: TestAccAPIGatewayDomainName_tags_IgnoreTags_Overlap_ResourceTag (70.18s)
--- PASS: TestAccAPIGatewayDomainNameDataSource_tags_IgnoreTags_Overlap_DefaultTag (246.75s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/apigateway	1210.848s
```

File: internal/service/apigateway/domain_name.go [hunk 1]
```diff
@@ -416,10 +472,43 @@ func resourceDomainNameDelete(ctx context.Context, d *schema.ResourceData, meta
 	return diags
 }
 
-func findDomainByName(ctx context.Context, conn *apigateway.Client, domainName string) (*apigateway.GetDomainNameOutput, error) {
+const domainNameResourceIDSeparator = "/"
```
Comment (jar-b) [inline internal/service/apigateway/domain_name.go:475]: Can a comma delimiter work here to keep in line the with standard `flex.[Expand|Flatten]ResourceID` helper functions?

https://github.com/hashicorp/terraform-provider-aws/blob/42a5eedc7a92d5e36f394b0231ee657c9b5fab84/internal/flex/flex.go#L251-L253

https://github.com/hashicorp/terraform-provider-aws/blob/42a5eedc7a92d5e36f394b0231ee657c9b5fab84/internal/flex/flex.go#L22-L25

**Review by jar-b — APPROVED** [review]
One suggestion on the new multi-part ID delimiter, but otherwise LGTM 👍

File: internal/service/apigateway/domain_name.go (hunk 1 above)
Comment (ewbankkit) [reply internal/service/apigateway/domain_name.go:475]: Yes, it would -- I'll change

**Comment by ewbankkit** [comment]
After changed of resource ID separator:

```console
% make testacc TESTARGS='-run=TestAccAPIGatewayDomainName_regionalCertificateARN\|TestAccAPIGatewayDomainName_updateIDFormat\|TestAccAPIGatewayDomainName_private$$' PKG=apigateway
make: Verifying source code with gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.23.3 test ./internal/service/apigateway/... -v -count 1 -parallel 20  -run=TestAccAPIGatewayDomainName_regionalCertificateARN\|TestAccAPIGatewayDomainName_updateIDFormat\|TestAccAPIGatewayDomainName_private$ -timeout 360m
2024/12/04 16:43:15 Initializing Terraform AWS Provider...
=== RUN   TestAccAPIGatewayDomainName_regionalCertificateARN
=== PAUSE TestAccAPIGatewayDomainName_regionalCertificateARN
=== RUN   TestAccAPIGatewayDomainName_private
=== PAUSE TestAccAPIGatewayDomainName_private
=== RUN   TestAccAPIGatewayDomainName_updateIDFormat
=== PAUSE TestAccAPIGatewayDomainName_updateIDFormat
=== CONT  TestAccAPIGatewayDomainName_regionalCertificateARN
=== CONT  TestAccAPIGatewayDomainName_updateIDFormat
=== CONT  TestAccAPIGatewayDomainName_private
--- PASS: TestAccAPIGatewayDomainName_regionalCertificateARN (20.69s)
--- PASS: TestAccAPIGatewayDomainName_private (62.89s)
--- PASS: TestAccAPIGatewayDomainName_updateIDFormat (145.97s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/apigateway	151.637s
```

**Review by jar-b — APPROVED** [review]

**Comment by ewbankkit** [comment]
@DrFaust92 Thanks for the contribution :tada: :clap:.

File: internal/service/apigateway/domain_name.go (hunk 1 above)
Comment (DrFaust92) [reply internal/service/apigateway/domain_name.go]: Wasnt familiar with this, will use it from now on

**Pull request merged**